### PR TITLE
Optimizing Metric Implementation

### PR DESCRIPTION
**Pull request by gongna-au**
…ect performance

**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #433 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/metrics/prometheus/prometheus.go [hunk 1]
```diff
@@ -343,6 +344,7 @@ func (p *Prometheus) HandlerFunc() ContextHandlerFunc {
 		if err2 == nil {
 			p.resSz.WithLabelValues(statusStr, method, url).Observe(float64(resSz))
 		}
+		p.Ppg.counter = p.Ppg.counter + 1
```
Comment (ma642) [inline pkg/metrics/prometheus/prometheus.go:355]: 这个有并发问题，没有加锁.

File: pkg/metrics/prometheus/prometheus.go [hunk 2]
```diff
@@ -19,7 +19,6 @@ package prometheus
```
Comment (baerwang) [inline pkg/metrics/prometheus/prometheus.go:19]: will file move pixiu/pkg/

File: pkg/metrics/prometheus/prometheus.go [hunk 3]
```diff
@@ -261,40 +263,41 @@ func (p *Prometheus) registerMetrics() {
...
-func (p *Prometheus) getMetrics() []byte {
+func (p *Prometheus) SetPushGateway() {
+	if p.Ppg.CounterPush {
+		p.startPushCounter()
+	}
+}
 
+func (p *Prometheus) getMetrics() []byte {
 	out := &bytes.Buffer{}
 	metricFamilies, _ := prometheus.DefaultGatherer.Gather()
```
Comment (baerwang) [inline pkg/metrics/prometheus/prometheus.go]: def new error field,use it

**Review by baerwang — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Review by mark4z — APPROVED** [review]

File: pixiu/pkg/prometheus/prometheus.go [hunk 4]
```diff
@@ -343,6 +349,10 @@ func (p *Prometheus) HandlerFunc() ContextHandlerFunc {
 		if err2 == nil {
 			p.resSz.WithLabelValues(statusStr, method, url).Observe(float64(resSz))
 		}
+		var mutex sync.RWMutex
+		mutex.Lock()
+		p.Ppg.counter = p.Ppg.counter + 1
+		mutex.Unlock()
 		p.SetPushGateway()
```
Comment (ma642) [inline pixiu/pkg/prometheus/prometheus.go:357]: 这里会造成并发问题。

**Pull request merged**